docs/NewTargetSpecificNode.md

```diff
@@ -0,0 +1,93 @@
+## New Target Specific Node
+
+
+This document describes how to add new target specific node to make further optimization on certain backend.
+
+### Graph Optimization
+
+Glow has two levels of IR. The high-level IR is a dataflow node-based graph. There are several strategies to make transformation on high-level IR to achieve better performance on certain backend. Both these strategies will introduce new target specific node.
+
+#### Operator Fusion
+
+There are mainly two advantanges of operator fusion:
+
+- Good locality and less memory access
+- Reduce kernel launch time in GPU and specialized accelerators
+
+In TensorRT, the convolution, bias and ReLU layers of various sizes can be combined into a single kernel called CBR. A simple analogy is making three separate trips to the supermarket to buy three items versus buying all three in a single trip.
+
+
+#### Data Layout Transformation
+
+Actually, a Tensor is a view of a block of memory. Besides a pointer to the memory, we also have to get some other descriptions of this block of memory, such as shape, stride and layout.
+
+Different layout leads to different implementation of the operator kernel on certain hardware. For example, `NCHW` is a common layout of the input tensor of convolution operator. In Intel CPU, the best performance is achieved after transforming layout `NCHW` to layout `nChw8c`. After Tensor Core was introduced to Volta GPU, layout `NCHW` is not more efficient than layout `NHWC`.
+
+We should take both fast operator kernel implementation and extra potential layout transformation into consideration to get better performance.
+
+### Steps
+
+There are mainly three steps to add new target specific node in Glow:
+
+1. Add a backend-specific Node `FusedAB` to `XSpecificNodes.h`, and a corresponding backend-specific Instruction `FusedAB` to `XSpecificInstrs.h`. Note that the `FusedABInst` needs to be marked with `autoIRGen()` so that the node is automatically IRGen'd to the instruction, as we currently do not support backend-specific IRGen.
+
+2. Add logic to `XBackend::transformPreLowering()` or `XBackend::transformPostLowering()` (or both) depending on when you want to do the transformation. This logic would look for the pattern of nodes you want to fuse (`A` with a single use by `B`), and replaces all uses of the result of B with the new backend-specific `FusedABNode`.
+
+3. Have your backend `X` implement `FusedABInst`. For example, for the OpenCL backend, this would mean adding a case to enqueue a kernel for the `FusedABInst` to `OpenCLFunction::execute()`, and then adding the corresponding kernel in `kernels.cl`.
+
+
+### Examples
+
+#### Operator Fusion for ReLU in CPU
+
+ReLU is max between zero and the input value. Glow lowers ReLU Node to two basic low-level linear algebra operator nodes, Splat Node and Max Node.
```

Comment:
Can you expand on this a little bit here? Just say something about how a backend may want to optimize this by not requiring the `SplatNode`, fusing the Splat back into the Max, and that the CPU backend does this (leading to the next paragraph).

You also could reference our [Stacking instruction optimization](https://github.com/pytorch/glow/blob/master/docs/Optimizations.md#set-of-supported-ir-optimizations) somewhere, which is basically fusion of data-parallel instructions, just to be clear that we do some of this at the kernel/instruction level. Maybe not here, but perhaps somewhere else, you can decide if it fits somewhere.

Reply:
Done